src/k8s/pkg/client/kubernetes/certificate_signing_request.go

```diff
@@ -0,0 +1,68 @@
+package kubernetes
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/canonical/k8s/pkg/log"
+	certificatesv1 "k8s.io/api/certificates/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// WatchCertificateSigningRequest watches a CertificateSigningRequest with the
+// given name and calls a verify function on each event.
+// WatchCertificateSigningRequest will continue watching the CSR until the
+// verify function returns true or an non-retriable error occurs.
+//
+// The verify function should return true if the CSR is valid and processing
+// should stop.
+// The verify function should return false if the CSR is not yet valid and
+// processing should continue.
+// The verify function should return an error if the CSR is in an invalid state
+// (e.g., failed or denied) or the issued certificate is invalid.
+func (c *Client) WatchCertificateSigningRequest(ctx context.Context, name string, verify func(csr *certificatesv1.CertificateSigningRequest) (bool, error)) error {
+	log := log.FromContext(ctx)
+	for {
+		w, err := c.CertificatesV1().CertificateSigningRequests().Watch(ctx, metav1.SingleObject(metav1.ObjectMeta{Name: name}))
+		if err != nil {
+			log.V(1).Info("Failed to watch CSR", "error", err)
+			continue
+		}
+		defer w.Stop()
+		watchClosed := false
```

Comment:
having an internal function that implements the loop body will also help you with the `watchClosed`, that would then be able to go away

Reply:
I think offloading the loop body to another function will result in a function that returns a `(bool, error)` since we need to distinguish between retryable and non-retryable errors.

Is this acceptable for an internal function, or would you suggest another approach for this refactor?

Reply:
if this is internal, and callers do not need to concern themselves with the details, that is acceptable